test/ardop/test_ARDOPC.c

```diff
@@ -0,0 +1,131 @@
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <setjmp.h>
+#include <string.h>
+#include <cmocka.h>
+
+#include "common/ARDOPC.h"
+
+extern void ASCIIto6Bit(const char *Padded, UCHAR *Compressed);
+extern void Bit6ToASCII(const UCHAR *Padded, UCHAR *UnCompressed);
+extern void DeCompressCallsign(const char *bytCallsign, char *returned, size_t returnlen);
+extern void CompressCallsign(const char *Callsign, UCHAR *Compressed);
+
+/*
+ * Check that the 8-character ASCII string `out` compresses and
+ * decompresses back into `expect`. The input must be an 8-character
+ * string. Unused characters should be padded with spaces (U+0020).
+ */
+static void assert_6bit_inout(const char* out, const char* expect) {
+	UCHAR compressed[6];
+	char decompressed[9];
+
+	ASCIIto6Bit(out, compressed);
+	Bit6ToASCII(compressed, decompressed);
+	decompressed[8] = '\0';
+	assert_string_equal(decompressed, expect);
+}
+
+/*
+ * Check that the given `call`sign (with optional SSID) has the given
+ * compressed `hexstr` representation in the ARDOP protocol. The given
+ * `hexstr` must be 12 characters of hexadecimal like "b908e1b2c010"
+ */
+static void assert_callsign_wireline_is(const char *call, const char *hexstr)
+{
+	UCHAR compressed[6];
+	char compressed_hex[2*sizeof(compressed) + 1];
+	compressed_hex[2 * sizeof(compressed)] = '\0';
+
+	// check string must be exactly 12 characters long
+	assert_int_equal(strlen(hexstr), 2 * sizeof(compressed));
+
+	CompressCallsign(call, compressed);
+
+	for (size_t i = 0; i < sizeof(compressed); ++i) {
+		snprintf(&compressed_hex[2*i], 3, "%02x", (unsigned int)compressed[i]);
+	}
+	assert_string_equal(hexstr, compressed_hex);
+}
+
+/*
+ * Verify that the given `call_in` callsign (with optional SSID) becomes
+ * `call_out` after a round-trip through the callsign compress/decompress
+ * algorithm. This verifies that the callsign can be sent and received.
+ */
+static void assert_callsign_inout(const char *call_in, const char *call_out)
+{
+	char out[11];
+	UCHAR compressed[6];
+
+	CompressCallsign(call_in, compressed);
+	DeCompressCallsign(compressed, out, sizeof(out));
+	assert_string_equal(call_out, out);
+}
+
+/* test string-packing compression and decompression */
+static void test_ascii_6bit(void **state)
+{
+	(void)state; /* unused */
+
+	assert_6bit_inout("        ", "        ");
+	assert_6bit_inout("AZ09-_  ", "AZ09-_  ");
+	assert_6bit_inout("A      Z", "A      Z");
+	assert_6bit_inout("HI\0\0\0\0\0U", "HI      ");
+	assert_6bit_inout("abcxyz  ", "ABCXYZ  ");
+}
+
+/* test callsign wireline representation */
+static void test_compress_callsign(void **state)
+{
+	(void)state; /* unused */
+
+	// six character + SSID values are taken from direct
+	// invocation of CompressCallsign() in ardopcf eab1f3165a30.
```

Comment:
Please add comment that all example callsigns tested here and in `test_decompress_callsign` were manually tested by sending an IDFrame whose contents were encoded using `CompressCallsign()` to ARDOP_WIN v 1.0.2.6 where the received callsign was read from the `Rcv Frame` field of the associated ardop gui.  All values matched the expected result except for "LONGCAL-15" which was displayed as "LONGCAL-1".  It is unknown at this time whether that exception actually indicates that ARDOP_WIN actually decoded it incorrectly, or whether the ardop gui simply failed to display it properly.